src/crab/type_domain.hpp

```diff
@@ -0,0 +1,357 @@
+// Copyright (c) Prevail Verifier contributors.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <unordered_map>
+
+#include "crab/cfg.hpp"
+
+namespace crab {
+
+enum class region {
+	T_CTX,
+	T_STACK,
+	T_PACKET,
+	T_SHARED
+};
+
+inline std::string get_region(const region& r) {
+    switch (r) {
+        case region::T_CTX:
+            return "ctx";
+        case region::T_STACK:
+            return "stack";
+        case region::T_PACKET:
+            return "packet";
+        case region::T_SHARED:
+            return "shared";
+        default:
+            return "undefined";
+    }
+}
+
+inline std::ostream& operator<<(std::ostream& o, const region& t) {
+    o << static_cast<std::underlying_type<region>::type>(t);
+    return o;
+}
+
+struct ptr_no_off_t {
+    region r;
+
+    ptr_no_off_t(region _r) : r(_r) {}
+    ptr_no_off_t(const ptr_no_off_t& p) : r(p.r) {}
+
+    friend bool operator==(const ptr_no_off_t& p1, const ptr_no_off_t& p2) {
+        return (p1.r == p2.r);
+    }
+
+    friend bool operator!=(const ptr_no_off_t& p1, const ptr_no_off_t& p2) {
+        return !(p2 == p1);
+    }
+
+    friend std::ostream& operator<<(std::ostream& o, const ptr_no_off_t& p) {
+        return o << "{" << get_region(p.r) << "}";
+    }
+};
+
+struct ptr_with_off_t {
+    region r;
+    int offset;
+
+    ptr_with_off_t(region _r, int _off) : r(_r), offset(_off) {}
+    ptr_with_off_t(const ptr_with_off_t& p) : r(p.r), offset(p.offset) {}
+
+    friend bool operator==(const ptr_with_off_t& p1, const ptr_with_off_t& p2) {
+        return (p1.r == p2.r && p1.offset == p2.offset);
+    }
+
+    friend bool operator!=(const ptr_with_off_t& p1, const ptr_with_off_t& p2) {
+        return !(p1 == p2);
+    }
+
+    friend std::ostream& operator<<(std::ostream& o, const ptr_with_off_t& p) {
+        return o << "{" << get_region(p.r) << ", " << p.offset << "}";
+    }
+};
+
+using ptr_t = std::variant<ptr_no_off_t, ptr_with_off_t>;
+
+
+struct reg_with_loc_t {
+    int r;
+    std::pair<label_t, uint32_t> loc;
+
+    reg_with_loc_t() : r(-1), loc(std::make_pair(label_t::entry, 0)) {}
```

Comment:
I would avoid having here a default constructor and then define `r` as an `uint8_t`.
Maybe you defined it so that it's easier to use in maps but think twice about it.

Reply:
Default constructor is added because I have an array of size 11 that store `reg_with_loc_t` objects. Initializing such an array would require the default constructor for `reg_with_loc_t`